File: algorithms/leetcode/happy_number.go
package main
import (
  "fmt"
  "math"
  )


func is_happy(n int) bool {
  m := make(map[int]bool) 

  for n > 1 {
    if _, ok := m[n]; ok {
      return false
    }
    m[n] = true
    var temp int = 0 
    for n > 0 {
      temp += int(math.Pow(float64(n % 10), float64(2)))
      n /= 10
    }
    n = temp
  }
  return true 
}

func main() {

  fmt.Println(is_happy(19))
  fmt.Println(is_happy(2))
}

File: algorithms/leetcode/happy_number.rs
use std::collections::HashSet;

fn isHappy(mut n: i32) -> bool {
    let mut s: HashSet<i32> = HashSet::new();

    while n > 1 {
        if s.contains(&n) {
            return false;
        }

        let mut temp = 0;
        s.insert(n);
        while n > 0 {
            temp += i32::pow(n % 10, 2);
            n /= 10;
        }
        n = temp;
    }
    return true;
}

fn main() {
    println!("{}", isHappy(19));
    println!("{}", isHappy(2));
}
